// @katalon/testops-commons/dist/src/TestOpsReportGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestOpsReportGenerator = void 0;
var fs_1 = require("fs");
var path = require("path");
var TestOpsReportGenerator = /** @class */ (function () {
    function TestOpsReportGenerator(testOpsConfig) {
        this.testOpsConfig = testOpsConfig;
    }
    TestOpsReportGenerator.prototype.writeExecution = function (execution) {
        this.write(execution, "execution.json");
    };
    TestOpsReportGenerator.prototype.writeMetaData = function (metadata) {
        this.write(metadata, "metadata.json");
    };
    TestOpsReportGenerator.prototype.writeTestResults = function (testResults) {
        this.write(testResults, "results.json");
    };
    TestOpsReportGenerator.prototype.writeTestSuites = function (testSuites) {
        this.write(testSuites, "suites.json");
    };
    TestOpsReportGenerator.prototype.write = function (object, fileName) {
        var reportFolder = this.testOpsConfig.reportFolder;
        var data = JSON.stringify(object);
        var filePath = path.resolve(reportFolder, fileName);
        fs_1.mkdirSync(reportFolder, { recursive: true });
        fs_1.writeFileSync(filePath, data);
    };
    return TestOpsReportGenerator;
}());
exports.TestOpsReportGenerator = TestOpsReportGenerator;
